### ISSUE TITLE
Settings page for Firefox port

### ISSUE THREAD

**Issue by nickserv** [wyantb/HashMask#40]

**Comment by nickserv** [comment]
This could be tricky. I've been looking around, and it seems like Firefox's Jetpack SDK doesn't let you have preferences unless you're using Firefox's built in JSON preferences. Even though it's annoying, I'd rather make a native version of our settings page (and not have a settings page) than have a settings button stuck somewhere it doesn't belong in the extension's UI (in other words, I want to make sure users can get to our settings the way they're used to, via about:addons).

If someone could help me double check to see if it's possible to use a web page for preferences using the new Firefox SDK, that would be great. I haven't seen any other extensions do it, maybe XUL ones though (and that would probably require massive amounts of porting).

**Comment by wyantb** [comment]
Do you mean that you only see the options button display a popup, like Greasemonkey?

Take a look at the extension Restartless Restart 8 (first one in my extension list that behaves like this).  That one uses a settings page, although it's still not like Chrome's page.

**Comment by nickserv** [comment]
Yeah, that's using Firefox's built in settings stuff from the SDK. I don't think you can actually control the HTML/CSS/JS of it at all. I think you can only give it data, input fields, and what you want fields to link to in terms of data storage. We'd have to scrap our options page for the Firefox port if we wanted to use this.

**Comment by nickserv** [comment]
Also, I don't think Greasemonkey uses this new SDK yet, they probably used XUL to show that popup window of settings.

**Comment by deriachai** [comment]
There is always the full size option menu in firefox, it is nothing like what we currently have though, it is in their own format, not a free form page.

**Comment by nickserv** [comment]
You mean the embedded settings in Firefox's addons UI? Yeah, that's what I meant. We'd have to change it into JSONey stuff and ditch our options page. Actually, I feel like this is our best option for Firefox, since it's the standard anyway. Should I get started on it? Feel free to try to make any last minute attempts at having a web page as an options page with this SDK. You can probably do it with XUL, but our extension uses the new SDK so I don't think we can use that.

**Comment by wyantb** [comment]
Assumed you already were.  I will make very clear if I start working on any FF stuff again for a bit.  For right now, Chrome browser_action for me.

**Comment by nickserv** [comment]
Oh alright. I'll grab this issue, then.

Edit: Derp, I'm already assigned.

**Comment by wyantb** [comment]
Yip.  I should note that we're going to have to do some heavy merging and thinking later.  All the differences for FF, and now the index page will be different than the settings page, and the settings for FF will be entirely different...

In any case, the entire FF branch is yours with impunity, and don't worry about changing that unless I let you know in advance.

**Comment by nickserv** [comment]
Hm, I'm thinking about what I should do with the options/demo page in the Firefox version. Should I completely remove it, and maybe link to the demo somewhere? Or should I keep it but strip it down even more so it can be used as merely a preview page?

**Comment by wyantb** [comment]
You can do whatever you want.  Most likely, that is completely and thoroughly useless, so you'll want to axe it.  Our existing index.html in gh-pages should work for most browsers anyway.

**Comment by nickserv** [comment]
I guess, but it would still be useful as a preview page. Doesn't our demo ignore settings when HashMask is installed? It would be really cool if it didn't.

**Comment by wyantb** [comment]
Yes it does (edit: it does ignore settings, that is).  If we wanted a preview page...Go like NoScript does, and link to an external site?  Which happens to be our index on gh-pages?  

I really see limited usefulness for that page in FF.  But again, FF is entirely you until it's done.  Stop thinking and just do it.  We can discuss once it's ready (AKA, we have options.  FF can't get an initial release until it at least has options, so someone could theoretically sync between all their different installs on different computers.)
